Source/WebCore/svg/graphics/filters/SVGFilterTransaction.h

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2023 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+class FilterEffect;
+
+struct SVGFilterTransactionItem {
+    unsigned index;
```

Comment:
Index into what? This isn't the index in this Vector<>, right?

It's a bit confusing to use these vectors for SVG filters, where the filter structure is a graph, and more confusing to have this standalone structure that contains indexes into something else.

Reply:
It is an index to the `SVGFilter::m_effects` which is a Vector of the unique effects in the `SVGFilter` graph. The order of the effects in WebProcess and GPUProcess is the same. This is why when a certain `FilterEffect` changes in WebProcess we lookup its index in `SVGFilter::m_effects` and send it with this index to GPUProcess in `SVGFilterTransaction`. Using this index, GPUProcess will replace this effect in `SVGFilter::m_effects` and delete its `FilterImage` along with the result of all the dependent `FilterEffects`.

Reply:
`SVGFilterExpression` represents the graph of the `SVGFilter`. It is a Vector of `SVGFilterExpressionTerm`. Each `SVGFilterExpressionTerm` has an index pointing to the `SVGFilter::m_effects`.

Reply:
Is there a way we can package up SVGFilterExpression and SVGFilterTransaction into one struct, so we don't have these dangling indexes?

Reply:
Yes we can send the whole SVGFilter to GPUP and merge the new effects with the old cached one. If an effect does not match its cached corresponding one, we replace the old with the new and we clear the result and all the dependent results of this effect.